### PR TITLE
Make nonzero -> real & !zero like new assumptions

### PR DESCRIPTION
**Pull request by debugger22**
@asmeurer @jksuom Please review.

### CODE REVIEW

**Comment by debugger22** [comment]
This is the new behavior:
```
In [1]: from sympy import *

In [2]: x = Symbol('x', zero=False)

In [3]: x.is_nonzero

In [4]: x = Symbol('x', zero=False, real=True)

In [5]: x.is_nonzero
Out[5]: True
```

File: sympy/core/power.py [hunk 1]
```diff
@@ -352,6 +352,15 @@ def _eval_is_zero(self):
...
                 elif (1 - abs(self.base)).is_negative:
                     return self.exp.is_negative
+        else:
+            # when self.base.is_zero is None
+            if self.exp.is_finite:
+                return None
+            elif self.exp.is_infinite:
+                if (1 - abs(self.base)).is_positive:
+                    return self.exp.is_positive
+                elif (1 - abs(self.base)).is_negative:
```
Comment (jksuom) [inline sympy/core/power.py]: Is this possible in case self.base.is_zero is None?

Comment (debugger22) [inline sympy/core/power.py]: I didn't put much thought on this part. Just put it here to not break anything.
It looks like it won't be necessary. Thanks!

**Comment by jksuom** [comment]
[This](https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/complexes.py#L347) line and  ``x.is_real`` [here](https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/trigonometric.py#L2602) will become redundant.

**Comment by debugger22** [comment]
This change probably introduces a difficulty where one needs to check if something is not zero.

Unlike new assumptions, where we can use `~Q.zero(x)`, we can't use `not x.is_zero` because of `None`. Hence, we can't just replace `x.is_nonzero` with `not x.is_zero`. I'll push a commit with the workaround, soon.

File: sympy/functions/elementary/complexes.py [hunk 2]
```diff
@@ -498,7 +497,9 @@ def _eval_is_nonzero(self):
         return self._args[0].is_nonzero
 
     def _eval_is_positive(self):
-        return self.is_nonzero
+        is_z = self.is_zero
```
Comment (debugger22) [inline sympy/functions/elementary/complexes.py]: I'm unable to figure out the implementation of this `self.is_zero`.

Comment (debugger22) [inline sympy/functions/elementary/complexes.py]: Or perhaps, it's not there.

**Comment by asmeurer** [comment]
Use `fuzzy_not` in the old assumptions. Or just check `x.is_zero is False`.

File: sympy/functions/elementary/trigonometric.py [hunk 3]
```diff
@@ -2599,7 +2601,7 @@ def eval(cls, y, x):
                     return -S.Pi/2
                 elif y.is_zero:
                     return S.NaN
-        if y.is_zero and x.is_real and x.is_nonzero:
+        if y.is_zero and fuzzy_not(x.is_zero):
```
Comment (aktech) [inline sympy/functions/elementary/trigonometric.py]: @debugger22 
`fuzzy_not` will not work here:
```python
In [18]: x = I

In [19]: y = S.Zero

In [20]: y.is_zero and x.is_real and x.is_nonzero
Out[20]: False

In [21]: y.is_zero and fuzzy_not(x.is_zero)
Out[21]: True
```

Comment (debugger22) [inline sympy/functions/elementary/trigonometric.py]: Thanks a lot!

Comment (debugger22) [inline sympy/functions/elementary/trigonometric.py]: This is logically wrong. I shouldn't have removed `x.is_real`.

Comment (aktech) [inline sympy/functions/elementary/trigonometric.py]: Indeed!

**Comment by debugger22** [comment]
This is ready now.

**Comment by asmeurer** [comment]
Great. This use of fuzzy_not shows how the new assumptions syntax will be more useful. `ask(~Q.zero(x))` is a lot easier to read than `fuzzy_not(x.is_zero)`, and a lot less prone to mistakes with three-value logic.

**Comment by asmeurer** [comment]
Looks like a test failed.

**Comment by debugger22** [comment]
This failure is intermittent in this branch.

```
rerun 3
======================================================================= test process starts ========================================================================
executable:         /usr/bin/python  (2.7.6-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        83647877
hash randomization: on (PYTHONHASHSEED=3774641915)

sympy/solvers/tests/test_inequalities.py[1] F                                                                                                                 [FAIL]

____________________________________________________________________________________________________________________________________________________________________
______________________________________________ sympy/solvers/tests/test_inequalities.py:test_reduce_abs_inequalities _______________________________________________
  File "/home/debugger22/workspace/sympy/sympy/solvers/tests/test_inequalities.py", line 187, in test_reduce_abs_inequalities
    raises(TypeError, lambda: reduce_inequalities(abs(nr - 5) < 3))
  File "sympy/utilities/pytest.py", line 81, in raises
    raise AssertionError("DID NOT RAISE")
AssertionError: DID NOT RAISE
```

**Comment by debugger22** [comment]
@asmeurer Sorry for the delay. I forgot about this one. It should be working now.

**Comment by asmeurer** [comment]
Why was that intermittent? Does `_args` have different ordering in different sessions?

**Comment by asmeurer** [comment]
At any rate, if tests are passing I am +1 on this.

**Pull request merged**